### ISSUE TITLE
feat(benchmark): Create reviewer agent benchmark framework

### ISSUE THREAD

**Issue by anderskev** [existential-birds/amelia#8]
## Summary

Create a benchmark framework to systematically evaluate and iterate on the Reviewer agent. This enables data-driven improvements to review quality by measuring precision, recall, and severity calibration against a ground-truth dataset of code changes.

## Problem

Currently, we have no way to:
- Measure if changes to the reviewer prompts improve or regress performance
- Track reviewer quality over time
- Identify specific categories where the reviewer struggles

## Current Reviewer Architecture

The `Reviewer` agent (`amelia/agents/reviewer.py`) uses a single unified agentic method:

### Input

- **`agentic_review(state, base_commit, profile, *, workflow_id)`**
  - `state: ImplementationState` - Contains issue, goal/plan_markdown, driver_session_id, execution context
  - `base_commit: str` - Git commit hash to diff against (reviewer fetches diff via git tools)
  - `profile: Profile` - Working directory, driver configuration
  - `workflow_id: str` - For event streaming and logging

### Output

- Returns tuple: `(ReviewResult, session_id: str | None)`
  - `ReviewResult` contains:
    - `reviewer_persona: str` (always "Agentic")
    - `approved: bool` - Whether review approved the changes
    - `comments: list[str]` - Actionable issues (max 20, filtered to exclude positive observations)
    - `severity: Severity` - Overall severity ("low", "medium", "high", "critical")

- Also produces `StructuredReviewResult` with detailed `ReviewItem` objects containing:
  - `number`, `title`, `file_path`, `line`
  - `severity`: "critical", "major", "minor" (beagle format, mapped to output severity)
  - `issue`, `why`, `fix`

### Key Behavior

- Auto-detects file types and technologies in the diff
- Dynamically loads appropriate beagle review skills
- Fetches diffs via git tools (avoids "Argument list too long" errors)
- Uses unified `AgenticMessage` stream abstraction
- Single persona: "Agentic" (no competitive mode - strategy field was removed)

### Severity Mapping

The reviewer uses two severity scales that get mapped:

```python
# ReviewItem (beagle format): "critical", "major", "minor"
# ReviewResult (output): "low", "medium", "high", "critical"

# Mapping in _parse_review_result():
severity_priority = {"critical": 3, "major": 2, "minor": 1}
severity_map = {
    3: "critical",  # beagle critical → amelia critical
    2: "high",      # beagle major → amelia high
    1: "medium",    # beagle minor → amelia medium
    0: "low",       # no issues
}
```

### Directory Structure

```
benchmarks/
├── reviewer/
│   ├── __init__.py
│   ├── models.py           # BenchmarkCase, BenchmarkResult Pydantic models
│   ├── runner.py           # BenchmarkRunner - executes cases against Reviewer
│   ├── evaluator.py        # Metrics calculation (accuracy, correlation, detection)
│   ├── detection.py        # Keyword/pattern-based issue detection matching
│   ├── report.py           # Markdown report generation
│   ├── fixtures.py         # Git repository fixture management
│   └── cases/
│       ├── __init__.py
│       ├── security/
│       │   ├── sql_injection.yaml
│       │   ├── xss.yaml
│       │   ├── command_injection.yaml
│       │   └── hardcoded_secrets.yaml
│       ├── performance/
│       │   ├── algorithmic_complexity.yaml
│       │   ├── memory_leaks.yaml
│       │   └── missing_caching.yaml
│       ├── correctness/
│       │   ├── logic_errors.yaml
│       │   ├── off_by_one.yaml
│       │   ├── null_handling.yaml
│       │   └── race_conditions.yaml
│       └── clean/
│           ├── good_implementations.yaml
│           └── valid_refactoring.yaml
```

### Data Models

```python
class BenchmarkCase(BaseModel):
    """A single benchmark test case."""
    id: str                           # e.g., "sec-001-sql-injection"
    name: str                         # Human-readable name
    category: Literal["security", "performance", "correctness", "clean"]
    difficulty: Literal["easy", "medium", "hard"]
    tags: list[str]                   # For filtering
    
    # Input context (for ImplementationState)
    issue_title: str
    issue_description: str
    
    # Git fixture approach (reviewer fetches diff via git tools)
    base_files: dict[str, str]        # filename -> content before changes
    changed_files: dict[str, str]     # filename -> content after changes
    
    # Ground truth
    expected_approved: bool
    expected_severity: Severity       # "low", "medium", "high", "critical"
    expected_issues: list[str]        # Issue types that should be detected


class BenchmarkResult(BaseModel):
    """Result of running a single benchmark case."""
    case_id: str
    actual_approved: bool
    actual_severity: Severity
    actual_comments: list[str]
    structured_items: list[ReviewItem] | None  # If structured review available
    
    # Computed metrics
    approval_correct: bool
    severity_match: bool
    detected_issues: list[str]        # Which expected issues were found
    missed_issues: list[str]          # Which expected issues were missed
    
    latency_ms: float
    token_usage: DriverUsage | None


class BenchmarkRun(BaseModel):
    """Results of a complete benchmark run."""
    run_id: str
    timestamp: datetime
    driver: str                       # Which driver was used (e.g., "api:openrouter")
    model: str                        # Model used for review
    results: list[BenchmarkResult]
    
    # Aggregate metrics
    approval_accuracy: float          # % cases with correct approval decision
    severity_correlation: float       # Spearman correlation
    issue_detection_recall: float     # % of expected issues detected
    false_positive_rate: float        # % of clean code flagged
```

### Git Fixture Approach

Since the reviewer fetches diffs via git tools rather than accepting raw diff strings, benchmarks need to create temporary git repositories:

```python
class GitFixture:
    """Creates temporary git repos for benchmark cases."""
    
    def __init__(self, case: BenchmarkCase):
        self.case = case
        self.repo_path: Path | None = None
    
    async def setup(self) -> tuple[Path, str]:
        """Create temp repo with base commit, return (path, base_commit_hash)."""
        # 1. Create temp directory
        # 2. git init
        # 3. Write base_files, commit as "base"
        # 4. Write changed_files, commit as "changes"
        # 5. Return repo path and base commit hash
    
    async def teardown(self) -> None:
        """Clean up temp repo."""
```

### Issue Detection Patterns

To evaluate whether the reviewer "caught" a specific issue type:

```python
ISSUE_DETECTION_PATTERNS: dict[str, list[str]] = {
    "sql_injection": [
        "sql injection", "parameterized query", "prepared statement",
        "sanitize", "escape", "string concatenation"
    ],
    "xss": [
        "cross-site scripting", "xss", "escape html", "sanitize input",
        "innerHTML", "user input"
    ],
    "command_injection": [
        "command injection", "shell injection", "subprocess",
        "os.system", "shell=True"
    ],
    "hardcoded_secret": [
        "hardcoded", "secret", "credential", "password", "api key",
        "environment variable"
    ],
    "algorithmic_complexity": [
        "O(n²)", "O(n^2)", "nested loop", "quadratic", "complexity",
        "optimize", "performance"
    ],
    # ... more patterns
}
```

### Metrics

| Metric | Description | Target |
|--------|-------------|--------|
| **Approval Accuracy** | % of cases where `approved` matches `expected_approved` | >90% |
| **Severity Correlation** | Spearman rank correlation between predicted and expected severity | >0.7 |
| **Issue Detection Recall** | % of `expected_issues` that appear in comments | >80% |
| **False Positive Rate** | % of `clean` cases incorrectly flagged with issues | <10% |
| **Mean Latency** | Average time per review | <30s |

### CLI Interface

```bash
# Run full benchmark
uv run amelia benchmark reviewer

# Filter by category
uv run amelia benchmark reviewer --category security

# Filter by tags
uv run amelia benchmark reviewer --tag sql --tag injection

# Compare two runs
uv run amelia benchmark reviewer --compare run-abc123 run-def456

# Output formats
uv run amelia benchmark reviewer --output results.json
uv run amelia benchmark reviewer --output report.md

# Use specific driver
uv run amelia benchmark reviewer --driver api:openrouter
```

### Example Benchmark Case (YAML)

```yaml
# benchmarks/reviewer/cases/security/sql_injection.yaml
- id: sec-001-sql-injection-basic
  name: Basic SQL injection in user lookup
  category: security
  difficulty: easy
  tags: [sql, injection, input-validation]
  
  issue_title: "Add user search endpoint"
  issue_description: |
    Implement GET /api/users/search?q=<query> to search users by name.
  
  base_files:
    api/users.py: |
      from fastapi import APIRouter
      
      router = APIRouter()
      
      @router.get("/")
      async def list_users():
          return await db.fetch_all("SELECT * FROM users")
  
  changed_files:
    api/users.py: |
      from fastapi import APIRouter
      
      router = APIRouter()
      
      @router.get("/")
      async def list_users():
          return await db.fetch_all("SELECT * FROM users")
      
      @router.get("/search")
      async def search_users(q: str):
          query = f"SELECT * FROM users WHERE name LIKE '%{q}%'"
          result = await db.execute(query)
          return result.fetchall()
  
  expected_approved: false
  expected_severity: critical
  expected_issues:
    - sql_injection

- id: sec-002-sql-injection-parameterized
  name: Proper parameterized query (clean)
  category: clean
  difficulty: easy
  tags: [sql, security-fixed]
  
  issue_title: "Add user search endpoint"
  issue_description: |
    Implement GET /api/users/search?q=<query> to search users by name.
  
  base_files:
    api/users.py: |
      from fastapi import APIRouter
      
      router = APIRouter()
      
      @router.get("/")
      async def list_users():
          return await db.fetch_all("SELECT * FROM users")
  
  changed_files:
    api/users.py: |
      from fastapi import APIRouter
      
      router = APIRouter()
      
      @router.get("/")
      async def list_users():
          return await db.fetch_all("SELECT * FROM users")
      
      @router.get("/search")
      async def search_users(q: str):
          query = "SELECT * FROM users WHERE name LIKE :pattern"
          result = await db.execute(query, {"pattern": f"%{q}%"})
          return result.fetchall()
  
  expected_approved: true
  expected_severity: low
  expected_issues: []
```

---

### Phase 0: Git Fixture Infrastructure
- [ ] Create `benchmarks/reviewer/fixtures.py` with `GitFixture` class
- [ ] Implement temp repo creation with base/changed commits
- [ ] Add cleanup logic for test repos
- [ ] Test fixture creation and diff generation

### Phase 1: Foundation
- [ ] Create `benchmarks/reviewer/models.py` with `BenchmarkCase`, `BenchmarkResult`, `BenchmarkRun`
- [ ] Create `benchmarks/reviewer/__init__.py` with YAML case loader
- [ ] Create `benchmarks/reviewer/runner.py` with `BenchmarkRunner` class
- [ ] Add basic test to verify runner executes cases

### Phase 2: Evaluation
- [ ] Create `benchmarks/reviewer/detection.py` with `ISSUE_DETECTION_PATTERNS`
- [ ] Create `benchmarks/reviewer/evaluator.py` with metric calculations
- [ ] Implement `approval_accuracy()` metric
- [ ] Implement `severity_correlation()` metric (scipy.stats.spearmanr)
- [ ] Implement `issue_detection_recall()` metric
- [ ] Implement `false_positive_rate()` metric

### Phase 3: Test Cases
- [ ] Create 5 security cases (SQL injection, XSS, command injection, hardcoded secrets)
- [ ] Create 5 performance cases (O(n²), memory, caching)
- [ ] Create 5 correctness cases (logic errors, off-by-one, null handling)
- [ ] Create 5 clean cases (good code that should be approved)
- [ ] Document ground truth methodology

### Phase 4: CLI & Reporting
- [ ] Add `benchmark` subcommand to Typer CLI (`amelia/main.py`)
- [ ] Create `benchmarks/reviewer/report.py` for markdown report generation
- [ ] Implement `--category` and `--tag` filtering
- [ ] Implement `--output` for JSON/Markdown export
- [ ] Implement `--compare` for run comparison

### Phase 5: Iteration Support
- [ ] Store benchmark runs in `benchmarks/reviewer/results/`
- [ ] Implement run history and trend visualization
- [ ] Add CI integration guidance (run benchmark on PR changes to reviewer)
- [ ] Document workflow for prompt iteration

---

## Success Criteria

1. Can run `uv run amelia benchmark reviewer` and get a scored result
2. Benchmark includes at least 20 diverse test cases across categories
3. Metrics clearly show whether reviewer changes improve or regress quality
4. Can compare two benchmark runs and see diff in metrics
5. Results are reproducible (same cases, same driver → similar scores)

## Future Enhancements

- **LLM-as-judge**: Use a separate LLM to evaluate comment quality semantically
- **Human validation**: Track which benchmark results humans agree/disagree with
- **Prompt optimization**: Use benchmark to A/B test different system prompts
- **Multi-model comparison**: Compare reviewer quality across different LLM backends

**Comment by anderskev** [comment]
Closing as **superseded by [daydream](https://github.com/existential-birds/daydream)**.

## Decision

We are moving code review out of Amelia entirely. The `Reviewer` agent (`amelia/agents/reviewer.py`) will be removed and reviews will be delegated to daydream. That makes a *reviewer-specific* benchmark framework inside Amelia the wrong place to build this — the agent it would measure is going away.

## Why daydream already covers the intent of this issue

The goal here was to (1) measure whether reviewer changes improve/regress quality, (2) track review quality over time, and (3) identify categories where the reviewer struggles. daydream is purpose-built around exactly that loop — it "produces structured training data from its own runs" and projects them into scored datasets:

- **Ground-truth scoring & metrics** — daydream's corpus pipeline (`daydream harvest` / `daydream/training/reward.py`) computes a composite intrinsic reward per run from per-finding verifier verdicts (`consistent`/`uncertain`/`contradicts`) and grounding rate, plus a posterior false-positive cost calibrated against maintainer accept/reject labels. This is the data-driven precision/recall/calibration measurement this issue asked for, at corpus scale rather than a fixed 20-case fixture set.
- **Quantitative eval** — `daydream/eval/analyzer.py` parses ATIF v1.6 trajectories and computes metrics across cost, tool usage, file coverage, finding quality, and grounding.
- **External benchmark targets** — daydream is benchmarked against [Martian's Code Review Bench](https://github.com/withmartian/code-review-benchmark) (Sentry, Grafana, Cal.com, Discourse, Keycloak) with explicit precision/F1/false-positive-per-run targets. See the [Milestone 1 training roadmap](https://github.com/existential-birds/daydream/issues/86) and [Milestone 2](https://github.com/existential-birds/daydream/issues/102).
- **Reproducibility** — `daydream build-corpus` writes a content-addressed `lineage.json` manifest with an `as_of` pin for byte-for-byte reproducible dataset snapshots, satisfying the "reproducible results" success criterion.

The git-fixture / YAML-case approach proposed here is effectively replaced by daydream's real-PR replay corpus and held-out benchmark, which exercises real review trajectories rather than synthetic single-file diffs.

If we later want Amelia-side regression gating, it should consume daydream's archive/index rather than re-implementing a parallel benchmark harness against an agent we're deleting.
